### PR TITLE
Apply redirect to signup

### PR DESCRIPTION
**Pull request by VictiniX888**
### Description

(_What have been done/changed_)

- Users visiting the apply-2023 page while not logged in will be redirected to the signup page
- Remember path of URL that user tried to visit so that we can redirect back to the exact path
- [Vincent] fixed up hackform: title on ethnicity changes, dropdown options on grad year redo, application status banner fix CSS, removed last two MLH questions 

### Testing plan

(_How will we test the changes requested on this PR?_)

- [ ] ...

### CODE REVIEW

File: libs/sso-client/src/lib/sso-client.ts [hunk 1]
```diff
@@ -8,75 +8,112 @@ import axios from 'axios';
...
             }
+          }
+        }
+
+        // Access token not found
+
+        // Redirect to show unauthorized if it is API endpoint
+        if (path.length >= 2 && path[1] === 'api') {
+          return NextResponse.redirect(
+            `${process.env.NEXT_PUBLIC_SSO_URL}/api/unauthorized`
```
Comment (snokpok) [inline libs/sso-client/src/lib/sso-client.ts:93]: nit: use `getEnv()`

File: libs/sso-client/src/lib/sso-client.ts [hunk 2]
```diff
@@ -203,3 +242,37 @@ async function verifyToken(access_token: string, refresh_token: string) {
...
   window.location.replace(`${process.env.NEXT_PUBLIC_SSO_URL}/logout`);
 }
+
+function generateRedirectUrl(
+  callbackUrl: string,
+  path: string,
+  redirect?: string
+): URL {
+  const redirectUrl = new URL(
+    redirect ?? `${process.env.NEXT_PUBLIC_SSO_URL}/login`
```
Comment (snokpok) [inline libs/sso-client/src/lib/sso-client.ts:252]: use `getEnv()`

File: libs/sso-client/src/lib/sso-client.ts [hunk 3]
```diff
@@ -8,75 +8,112 @@ import axios from 'axios';
...
+ * @param redirectUrl URL to redirect unauthorized users to
+ * @param exhaustive Whether the middleware should terminate or propagate. Default true (terminate)
  * @returns NextJS middleware function
  */
 export const middlewareHandler =
-  (callbackUrl: string) => async (request: NextRequest) => {
+  (
+    callbackUrl: string,
+    guardPaths?: string[],
+    redirectUrl?: string,
```
Comment (snokpok) [inline libs/sso-client/src/lib/sso-client.ts:20]: nit: -> `unauthorizedRedirectUrl`

**Review by snokpok — CHANGES_REQUESTED** [review]
minor changes

File: libs/sso-client/src/lib/sso-client.ts (hunk 1 above)
Comment (VictiniX888) [reply libs/sso-client/src/lib/sso-client.ts]: Done!

File: libs/sso-client/src/lib/sso-client.ts (hunk 2 above)
Comment (VictiniX888) [reply libs/sso-client/src/lib/sso-client.ts]: Done!

File: libs/sso-client/src/lib/sso-client.ts (hunk 3 above)
Comment (VictiniX888) [reply libs/sso-client/src/lib/sso-client.ts]: Done!

**Review by snokpok — APPROVED** [review]
lgtm

**Pull request merged**